### PR TITLE
JS: add js/session-fixation query

### PR DESCRIPTION
**Pull request by erik-krogh**
Inspired by [this csharp query](https://github.com/github/codeql/blob/e2cb53c65f804bd90bdce9313165d328cec46e70/csharp/ql/src/Security%20Features/CWE-384/AbandonSession.ql).  (Which has no results on all of lgtm.com). 

[An evaluation looks ok-ish](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/cwe384__default__CustomSuite/reports).  
The new results appear to be TPs, but I'm not sure if they are worth fixing.  
So I set the severity to `warning` and the precision to `medium`, so that the alerts are not shown by default.

### CODE REVIEW

File: javascript/ql/src/Security/CWE-384/SessionFixation.ql [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        .getAPropertyRead("session")
+        .getAPropertyWrite()
+  )
+  or
+  // or an authentication method is used (e.g. `passport.authenticate`)
+  setup.getARouteHandler().(DataFlow::CallNode).getCalleeName() = "authenticate"
+}
+
+/**
+ * Holds if `handler` is regenerates it's session using `req.session.regenerate`.
```
Comment (asgerf) [inline javascript/ql/src/Security/CWE-384/SessionFixation.ql:39]:
```suggestion
 * Holds if `handler` regenerates its session using `req.session.regenerate`.
```

**Review by asgerf — COMMENTED** [review]
After briefly reading up on session fixation, my initial impression is that `express-session` in its default configuration is practically immune, and the attack is only viable if the site already has an XSS vulnerability.

With the `warning/medium` classification I think that's fine, but I'm curious if I missed anything?

Otherwise LGTM

**Comment by erik-krogh** [comment]
> After briefly reading up on session fixation, my initial impression is that `express-session` in its default configuration is practically immune, and the attack is only viable if the site already has an XSS vulnerability.
> 
> With the `warning/medium` classification I think that's fine, but I'm curious if I missed anything?

I don't think you did.  
But [some developers care about it anyway](https://github.com/jaredhanson/passport/issues/192).

**Review by asgerf — DISMISSED** [review]

File: javascript/ql/src/Security/CWE-384/SessionFixation.qhelp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+</p>
+<sample src="examples/SessionFixationFixed.js" />
+
+</example>
+<references>
+<li>
+  OWASP: <a href="https://www.owasp.org/index.php/Session_fixation">Session fixation</a>
+</li>
+<li>
+  Stackoverflow: <a href="https://stackoverflow.com/questions/22209354/creating-a-new-session-after-authentication-with-passport/30468384#30468384">Creating a new session after authentication with Passport</a>
```
Comment (ethanpalm) [inline javascript/ql/src/Security/CWE-384/SessionFixation.qhelp:42]:
```suggestion
  Stack Overflow: <a href="https://stackoverflow.com/questions/22209354/creating-a-new-session-after-authentication-with-passport/30468384#30468384">Creating a new session after authentication with Passport</a>
```

**Review by ethanpalm — APPROVED** [review]
This looks good from a docs perspective with just one tiny suggestion 👍🏼

**Review by ethanpalm — APPROVED** [review]
Approving again since my previous review was dismissed as stale.

Looks good from docs 🚀

**Comment by jaredhanson** [comment]
Thanks for all your work on security analysis!  I just wanted to update here that the latest version of `passport`, 0.6.0, is now actively defending against session fixation attacks by default.  Sessions are regenerated internally, and there's no longer any need to do it at the application-level. The [announcement](https://medium.com/passportjs/fixing-session-fixation-b2b68619c51d) has more info.  Let me know if you want any further details.

**Comment by erik-krogh** [comment]
> Thanks for all your work on security analysis! I just wanted to update here that the latest version of `passport`, 0.6.0, is now actively defending against session fixation attacks by default. Sessions are regenerated internally, and there's no longer any need to do it at the application-level. The [announcement](https://medium.com/passportjs/fixing-session-fixation-b2b68619c51d) has more info. Let me know if you want any further details.

Thanks for the heads up. I'll [remove passport from our query](https://github.com/github/codeql/pull/9261).  

If you [create a security advisory](https://github.com/jaredhanson/passport/security/advisories) (and possibly request a CVE), then more people will probably upgrade to the new version.

**Pull request merged**